userd/autostart.go

```diff
@@ -0,0 +1,177 @@
+		}
+		break
+	}
+
+	execCmd = strings.TrimSpace(execCmd)
+	if execCmd == "" {
+		return "", fmt.Errorf("Exec not found or invalid")
+	}
+	return execCmd, nil
+}
+
+func getCurrentSnapInfo(snapName string) (*snap.Info, error) {
+	curFn := filepath.Join(dirs.SnapMountDir, snapName, "current")
+	realFn, err := os.Readlink(curFn)
+	if err != nil {
+		return nil, fmt.Errorf("cannot find current revision for snap %s: %s", snapName, err)
+	}
+	rev := filepath.Base(realFn)
+	revision, err := snap.ParseRevision(rev)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read revision %s: %s", rev, err)
+	}
+
+	return snap.ReadInfo(snapName, &snap.SideInfo{Revision: revision})
+}
+
+func tryAutostartApp(snapName, desktopFilePath string) (*exec.Cmd, error) {
+	desktopFile := filepath.Base(desktopFilePath)
+
+	info, err := getCurrentSnapInfo(snapName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to obtain snap information for snap %q: %v", snapName, err)
+	}
+
+	var app *snap.AppInfo
+	for _, candidate := range info.Apps {
+		if candidate.Autostart == desktopFile {
+			app = candidate
+			break
+		}
+	}
+	if app == nil {
+		return nil, fmt.Errorf("failed to match desktop file %q with snap %q applications", desktopFile, snapName)
+	}
+
+	content, err := ioutil.ReadFile(desktopFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	command, err := findExec(content)
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine startup command: %v", err)
+	}
+	logger.Debugf("exec line: %v", command)
+
+	split, err := shlex.Split(command)
+	if err != nil {
+		return nil, fmt.Errorf("invalid application startup command: %v", err)
+	}
+
+	cmd := exec.Command(app.WrapperPath(), split[1:]...)
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = os.Stdout
+	if err := cmd.Start(); err != nil {
+		return nil, fmt.Errorf("failed to autostart %q: %v", desktopFile, err)
+	}
+	return cmd, nil
+}
+
+type failedAutostartError map[string]error
+
+func (f failedAutostartError) Error() string {
+	var out bytes.Buffer
+	for app, err := range f {
+		fmt.Fprintf(&out, "- %q: %v\n", app, err)
+	}
+	return out.String()
+}
+
+var userCurrent = user.Current
+
+// AutostartSessionApps starts applications which have placed their desktop
+// files in $SNAP_USER_DATA/.config/autostart
+func AutostartSessionApps() error {
+	usr, err := userCurrent()
+	if err != nil {
+		return err
+	}
+
+	usrSnapDir := filepath.Join(usr.HomeDir, "snap")
+
+	glob := filepath.Join(usrSnapDir, "*/current/.config/autostart/*.desktop")
+	matches, err := filepath.Glob(glob)
+	if err != nil {
+		return err
+	}
+
+	failedApps := make(failedAutostartError)
+	for _, desktopFilePath := range matches {
+		desktopFile := filepath.Base(desktopFilePath)
+		logger.Debugf("autostart desktop file %v", desktopFile)
+
+		// /home/foo/snap/some-snap/current/.config/autostart/some-app.desktop ->
+		//    some-snap/current/.config/autostart/some-app.desktop
+		noHomePrefix := strings.TrimPrefix(desktopFilePath, usrSnapDir+"/")
```

Comment:
Is this just `filepath.Rel` done by hand?

Reply:
Yes and no. Skips all the fuss that `filepath.Rel()` does because the input is known to contain the prefix already.